docs/src/constraints.md

```diff
@@ -1,5 +1,514 @@
+types of functions and sets that JuMP knows about as needed. You can read more
+details about this *function-in-set* concept in the MathOptInterface
+documentation.
+
+!!! note
+    Throughout this page (and these docs), we use `MOI` as a shorthand for the
+    `MathOptInterface` module. This can be created by including the following
+    lines after `using JuMP` in your code.
+    ```julia
+    using MathOptInterface
+    const MOI = MathOptInterface
+    ```
+
+## The `@constraint` macro
+
+Constraints are added to a JuMP model using the [`@constraint`](@ref) macro. It
+is similar to the [`@variable`](@ref) macro. Here is an example of how to add
+the constraint ``2x \le 1`` to a JuMP model:
+```jldoctest con1; setup = :(model = Model(); @variable(model, x))
+julia> @constraint(model, con, 2x <= 1)
+con : 2 x <= 1.0
+```
+Wasn't that easy! Let's unpack what happened, because just like
+[`@variable`](@ref) there are a few subtle things going on.
+ 1. The mathematical constraint ``2x \le 1`` was added to the model.
+ 2. A JuMP called `con` was created that acts as a reference to the
+    constraint.
+ 3. JuMP set the *name* of the constraint to `"con"`.
+
+Just like the Julia variables created in [`@variable`](@ref), `con` can be bound
+to a different value. For example:
+```jldoctest con1
+julia> con
+con : 2 x <= 1.0
+
+julia> con = 1
+1
+
+julia> con
+1
+```
+However, the reference can be retrieved by querying the model using the symbolic
+name:
+```jldoctest con1
+julia> con = model[:con]
+con : 2 x <= 1.0
+
+julia> con
+con : 2 x <= 1.0
+```
+Because the named variables and constraints are stored in the same namespace,
+creating a constraint with the same name as a variable or an existing constraint
+will result in an error. To overcome this limitation, it is possible to create
+anonymous constraints, just like it is possible to create
+[Anonymous JuMP variables](@ref). This is done by dropping the second argument
+to [`@constraint`](@ref):
+```jldoctest con1
+julia> con = @constraint(model, 2x <= 1)
+2 x <= 1.0
+```
+
+It is also possible use different comparison operators (e.g., `>=` and `==`) to
+create the following types of constraints:
+```jldoctest con1
+julia> @constraint(model, 2x >= 1)
+2 x >= 1.0
+
+julia> @constraint(model, 2x == 1)
+2 x = 1.0
+
+julia> @constraint(model, 1 <= 2x <= 3)
+2 x ∈ [1.0, 3.0]
+```
+
+Note that JuMP normalizes the constraints given by the user by moving all of the
+terms containing variables to the left-hand side, and all of the constant terms
+to the right-hand side. Thus, we get:
+```jldoctest; setup=:(model=Model(); @variable(model, x))
+julia> @constraint(model, 2x + 1 <= 4x + 4)
+-2 x <= 3.0
+```
+
+## [Duality](@id constraint_duality)
+
+JuMP adopts the notion of duality from MathOptInterface. Roughly speaking, the
+dual variable associated with a constraint can be viewed as the decrease in the
+value of the objective given an infinitesimal relaxation of the constraint. For
```

Comment:
This relaxation logic corresponds to how I understand `shadow_price`. The only difference between this text and the definition of `shadow_price` is that "the decrease in the objective" becomes "the improvement in the objective". Was that intended?

The relaxation argument also doesn't extend to conic constraints (how do you relax membership in a cone?). The heuristic that we should explain is that a feasible dual on a `>=` constraint is nonnegative and a feasible dual on a `<=` constraint is nonpositive. That should be enough for readers familiar with LP to figure it out.